ensime-sbt.el

```diff
@@ -88,26 +89,82 @@
   (interactive)
   (sbt-command "clean"))
 
+(defun ensime-sbt-do-gen-ensime ()
+  (interactive)
+  (sbt-command "gen-ensime"))
+
 (defun ensime-sbt-do-package ()
   (interactive)
   (sbt-command "package"))
 
-(defun ensime-sbt-do-test ()
+(defun ensime-sbt-find-subproject (file-name source-set config)
+  (when (and config source-set)
+    (let* ((subprojects (plist-get config :subprojects))
+	   (matches-subproject-dir? (lambda (dir)
+				      (string-match-p dir file-name)))
+	   (find-subproject (lambda (sp)
+			      (-any matches-subproject-dir? (plist-get sp :source-roots)))))
+      (-> (-find find-subproject subprojects)
+	  (plist-get :name)))))
+
+(defun ensime-sbt-test-dwim (command)
+  (let* ((file-name (or buffer-file-name default-directory))
+	 (source-set (cond
+		      ((string-match-p "src/test" file-name) "")
+		      ((string-match-p "src/it" file-name) "it:")
+		      ((string-match-p "src/fun" file-name) "fun:")))
+	 (config (if (ensime-connected-p)
+		     (ensime-config)
+		   (let ((f (ensime-config-find)))
+		     (when f (ensime-config-load f)))))
+	 (subproject (ensime-sbt-find-subproject file-name source-set config)))
+    (if (and source-set subproject)
+	(sbt-command (concat subproject "/" source-set command))
+      (hydra-ensime-sbt-test/body))))
+
+(defun ensime-sbt-do-test-dwim ()
   (interactive)
-  (sbt-command "test"))
+  (ensime-sbt-test-dwim "test"))
 
-(defun ensime-sbt-do-test-quick ()
+(defun ensime-sbt-do-test-quick-dwim ()
   (interactive)
-  (sbt-command "testQuick"))
+  (ensime-sbt-test-dwim "testQuick"))
 
-(defun ensime-sbt-do-test-only ()
+(defun ensime-sbt-do-test-only-dwim ()
   (interactive)
   (let* ((impl-class
-            (or (ensime-top-level-class-closest-to-point)
-                (return (message "Could not find top-level class"))))
+	  (or (ensime-top-level-class-closest-to-point)
+	      (return (message "Could not find top-level class"))))
 	 (cleaned-class (replace-regexp-in-string "<empty>\\." "" impl-class))
 	 (command (concat "test-only" " " cleaned-class)))
-    (sbt-command command)))
+    (ensime-sbt-test-dwim command)))
+
+(defun ensime-guess-project-dir ()
+  (let ((ensime-file (ensime-config-find-file
```

Comment:
I've been doing some hacking in this area... there is a much better way to do this.

```elisp
  (let* ((connection (ensime-connection))
         (config (ensime-config connection))
      ...)
```